Tactics/Reflect/ProgramInfo.lean

```diff
@@ -17,81 +17,318 @@ Furthermore, we define a persistent env extension to store `ProgramInfo` in.
+        return val
+
+/-- Return `InstInfo.instSemantics?` from the state if it is `some _`,
+or use `f` to compute the relevant expressions if they are missing -/
+def getInstSemantics (f : Unit → InstInfoT m (Expr × Expr × Expr)) :
+    InstInfoT m (Expr × Expr × Expr) := do
+  let info ← get
+  match info.instSemantics? with
+    | some val => return val
+    | none =>
+        let val ← f ()
+        set {info with instSemantics? := some val}
+        return val
 
-def ProgramInfo.getRawInstrAt? (pi : ProgramInfo) (addr : BitVec 64) :
+end InstInfoT
+
+--------------------------------------------------------------------------------
+
+namespace ProgramInfo
+
+/-- The expression `mkConst pi.name`,
+i.e., an expression of this program referred to by name -/
+def expr (pi : ProgramInfo) : Expr := mkConst pi.name
+
+def getInstInfoAt? (pi : ProgramInfo) (addr : BitVec 64) :
+    Option InstInfo :=
+  pi.instructions.find? addr
+
+def getRawInstAt? (pi : ProgramInfo) (addr : BitVec 64) :
     Option (BitVec 32) :=
-  pi.rawProgram.find? addr
+  (·.rawInst) <$> pi.getInstInfoAt? addr
+
+-- TODO: this instance could be upstreamed (after cleaning it up)
+instance [BEq α] [Hashable α] : ForIn m (HashMap α β) (α × β) where
+  forIn map acc f := do
+    let f := fun (acc : ForInStep _) key val => do
+      match acc with
+        | .yield acc  => f ⟨key, val⟩ acc
+        | .done _     => return acc
+    match ← map.foldM f (ForInStep.yield acc) with
+     | .done x | .yield x => return x
 
-/-- Given an `Expr` of type `Program`, generate the basic `ProgramInfo` -/
-partial def ProgramInfo.generateFromExpr (e : Expr) : MetaM ProgramInfo := do
+/-! ## ProgramInfo Generation -/
+
+/-- Given the name and defining expression of a `Program`,
+generate the basic `ProgramInfo` -/
+partial def generateFromExpr (name : Name) (e : Expr) : MetaM ProgramInfo := do
+  trace[ProgramInfo] "Generating program info for `{name}` from definition:\n\t{e}"
   let type ← inferType e
   if !(←isDefEq type (mkConst ``Program)) then
     throwError "type mismatch: {e} {← mkHasTypeButIsExpectedMsg type (mkConst ``Program)}"
 
-  let rec go (rawProgram : HashMap _ _) (e : Expr) : MetaM (HashMap _ _) := do
+  let rec go (instructions : HashMap _ _) (e : Expr) : MetaM (HashMap _ _) := do
     let e ← whnfD e
     match_expr e with
     | List.cons _ hd tl => do
+        trace[ProgramInfo] "found address/instruction pair: {hd}"
+
         let hd' ← reduce hd
         let_expr Prod.mk _ _ addr inst := hd'
           | throwError "expected `{hd}` to reduce to an application of `Prod.mk`, found:\n\t{hd'}"
 
-        let addr ← reflectBitVecLiteral 64 addr
-        let inst ← reflectBitVecLiteral 32 inst
+        let addr ← reflectBitVecLiteral 64 (← instantiateMVars addr)
+        let inst ← reflectBitVecLiteral 32 (← instantiateMVars inst)
 
-        let rawProgram := rawProgram.insert addr inst
+        let rawProgram := instructions.insert addr ⟨inst, none, none⟩
```

Comment:
```suggestion
        let rawProgram := instructions.insert addr (InstInfo.ofRawInst inst)
```

and make a function that builds an `InstInfo`?

Reply:
I've made the other field optional, so this now reads `{ rawInst }`, does that spark joy?